Updated Azure Stack Bridge API Descriptions

lebarano opened the pull request:
### PR Description

Added missing Azure RP and Azure Stack Bridge API definitions

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

lebarano committed:
Aded missing APIs for RP

lebarano committed:
Added DownloadedProductPut operation

lebarano committed:
Added Configuration call description

lebarano committed:
Fixed configuration url structure

AutorestCI commented:
# Automation for azure-sdk-for-python

lebarano commented:
@bganapa, @deathly809, @martabarillas ,
Review please :)

AutorestCI commented:
# Automation for azure-sdk-for-go

martabarillas commented on specification/azsadmin/resource-manager/azurebridge/Microsoft.AzureBridge.Admin/preview/2016-01-01/AzureBridge.json:
```diff
@@ -266,6 +270,51 @@
+    "CompatibilityIssue": {
+      "description": "Compatibility issue",
+      "type": "string",
+      "enum": [
```
No unknown option?

lebarano replied:
Yep. If there are no known issues, product will be marked as compatible :)
And when it is - this array is empty

azuresdkci commented:
Can one of the admins verify this patch?

ryansbenson commented on specification/azsadmin/resource-manager/azurebridge/Microsoft.AzureBridge.Admin/preview/2016-01-01/DownloadedProduct.json:
```diff
@@ -154,6 +154,53 @@
+        ],
+        "responses": {
+          "200": {
+            "description": "OK"
```
You should declare the schema of the response (and it should be the same as the request with generated fields marked as readonly)

            "schema": {
              "$ref": "#/definitions/DownloadedProductPutResource"
            }

lebarano replied:
Updated the API

ryansbenson commented on specification/azsadmin/resource-manager/azurebridge/Microsoft.AzureBridge.Admin/preview/2016-01-01/DownloadedProduct.json:
```diff
@@ -154,6 +154,53 @@
+            "$ref": "AzureBridge.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/DownloadedProductPutParameter"
```
Can you share a common type definition with Get? DownloadedProductResource

ARM resources should be put-able with the response from the get request.

lebarano replied:
Updated the API

lebarano commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}/providers/Microsoft.AzureStack/registrations/{registrationName}/products/{productName}/GetProduct": {
```
We need this request to be POST, to add Configuration body
And RP SDK that we use for routing and underlying AzureStack configurations prohibit "Post" calls to "GetResource"
So that's a hack-around limitations we have on system

bganapa replied:
@ryansbenson We intend to deprecate the GET call on the product and replace it with this new POST call GetProducts, Can we simply delete the existing GET Product call? For this swagger , we have not published any SDK or powershell/CLI yet, so the impact is minimal . I am inclined to delete the existing GET call to remove the confusion. What do you suggest?

lebarano replied:
We discussed internally with team - for now we shouldn't delete GET requests entirely. In time it will most likely transform to behave differently from POST.

lebarano commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
         }
       }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}/providers/Microsoft.AzureStack/registrations/{registrationName}/products/_all/GetProducts": {
```
We need this request to be POST, to add Configuration body
And RP SDK that we use for routing and underlying AzureStack configurations prohibit "Post" calls to "GetResource"
So that's a hack-around limitations we have on system

bganapa replied:
@ryansbenson  We intend to deprecate the GET call on the product and replace it with this new POST call GetProducts, Can we simply delete the existing GET Product call? For this swagger , we have not published any SDK or powershell/CLI yet, so the impact is minimal . I am inclined to delete the existing GET call to remove the confusion. What do you suggest?

lebarano replied:
We discussed internally with team - for now we shouldn't delete GET requests entirely. In time it will most likely transform to behave differently from POST.

lebarano commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
+            "$ref": "AzureStack.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "#/parameters/MarketplaceProductLogUpdate"
```
This logs one action that was made on product
We want to use it to have a more detailed latest status for each product. It's designed on AzureStack side to send update on every action, and not batch them (just in case something goes wrong). So no reason to transform it into array.

Also, if we need to aggregate them - we can aggregate directly, without sending arrays

lebarano commented on specification/azsadmin/resource-manager/azurebridge/Microsoft.AzureBridge.Admin/preview/2016-01-01/AzureBridge.json:
```diff
@@ -207,6 +207,10 @@
           "description": "the product properties. At the moment only VirtualMachineProductProperties is allowed.",
           "$ref": "#/definitions/ProductProperties"
+        },
+        "compatibility": {
```
We're not likely to make any breaking changes at this stage, but still keeping APIs as preview for now.

lebarano commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -477,6 +667,106 @@
+          "type": "string",
+          "readOnly": true
+        },
+        "subscriptionId": {
```
They are part of "Resource", but not product itself. ProductID is defined by publisher and doesn't contain any AzureStack-specific information in it.
So those are not duplicates

lebarano commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroup}/providers/Microsoft.AzureStack/registrations/{registrationName}/products/{productName}/uploadProductLog": {
```
Sure there is. It's not defined within this file, because system that contains it is not accessible by customers (ACIS operations can access it), and is not described here at all.
Customers and RPs don't need access to that log, it's for our internal use only

lebarano commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -408,9 +549,58 @@
           "description": "Additional properties for the product.",
           "$ref": "#/definitions/ProductProperties"
+        },
+        "compatibility": {
```
we're adding this field for POST calls only (which we just invented), and keeping old API the same.
In future we will deprecate old API and keep only new one.

Should I create a different product description for this case, or is it fine to keep it like this?
New one will contain all the same stuff, just this one field added

bganapa commented:
This is being internally reviewed, please wait for our approval for merge

lebarano committed:
Updated APIs to comply with standard requirements

Added results to Product/download and DownloadedProduct/put calls
Allowed results of DownloadedProduct/get to be passed into DownloaedProduct/put call

lebarano committed:
Styling changes

Updated IdentitySystem to be enum
Updated MarketplaceProductLogUpdate parameter name

bganapa commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
+          "Product"
+        ],
+        "description": "Returns a list of products.",
+        "operationId": "GetProducts_List",
```
rename to Products_GetProducts ? sounds awful but we have the api like that!!

lebarano replied:
If I rename it to Products_GetProducts - i receive a couple more warnings by "autorest" tool. It actually forces this name if I want to avoid warnings.

I still can rename and live with the warnings, though, Should I?

jhendrixMSFT replied:
The convention is `Noun_Verb` where `Noun` is plural.  Since this returns a list of products I think the name should be `Products_List`

lebarano replied:
Again. If I rename it, I get autorest warning:
WARNING (PostOperationIdContainsUrlVerb/R2066/SDKViolation): OperationId should contain the verb: 'getproducts' in:'Products_List'. Consider updating the operationId

Should I ignore it and change the name?

jhendrixMSFT replied:
We can leave it as `Products_GetProducts`, it just means that for method group `Products` (i.e. IProducts) there will be two methods, GetProduct and GetProducts.

lebarano replied:
renamed to Products_GetProducts

bganapa commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
+          "Product"
+        ],
+        "description": "Returns the specified product.",
+        "operationId": "GetProduct",
```
rename to Products_GetProduct?

lebarano replied:
If I rename it to Products_GetProduct - i receive a couple more warnings by "autorest" tool. It actually forces this name if I want to avoid warnings.

I still can rename and live with the warnings, though, Should I?

jhendrixMSFT replied:
`Products_Get`

bganapa replied:
Products_get is already taken for the GET api . This is post call. I think you want to rename it to Products_GetProduct.. I think autorest SDK generation tool expects the form of OperationGroup_Operation @jhendrixMSFT can you pls confirm?

jhendrixMSFT replied:
If you name them both `Products_GetProduct` the code generators will apply their own rules to disambiguate the names which is undesirable.  I agree this one should be `Products_GetProduct`, let's rename the other one.

lebarano replied:
renamed to Products_GetProduct

jhendrixMSFT commented on specification/azurestack/resource-manager/Microsoft.AzureStack/stable/2017-06-01/Product.json:
```diff
@@ -143,6 +143,147 @@
+          "Product"
+        ],
+        "description": "Returns the specified product.",
+        "operationId": "Uploadproductlog",
```
`ProductLogs_Upload`

lebarano replied:
It uploads single log. Cannot update lists of logs and doesn't deal with multiple logs ever.
Why plural?

bganapa replied:
Products_UploadLog ?

lebarano replied:
renamed to Products_UploadLog

jhendrixMSFT commented on specification/azsadmin/resource-manager/azurebridge/Microsoft.AzureBridge.Admin/preview/2016-01-01/Configuration.json:
```diff
@@ -0,0 +1,96 @@
+        "tags": [
+          "Configuration"
+        ],
+        "operationId": "Configuration_Get",
```
`Configurations_Get`

lebarano replied:
It gets a single configuration. Not list of them
Why would it be "Configurations"?

bganapa replied:
I think the plural is the typical norm for all the ARM resources

jhendrixMSFT replied:
The noun defines the operation group name, not the thing(s) the verb returns.  It's plural so that the response model name can be singular else the names collide.  In this case there is no collision.  Is the configuration really singular?  If so then I'm fine with leaving it as singular assuming there isn't another operation group with the same name.

lebarano replied:
there are no other configuration getters and yes, it's really singular)

bganapa commented:
@jhendrixMSFT  Could you please help rerunning the validation? I clicked rerun, but it doesnt seem to be doing the run.

bganapa commented:
@lebarano There are few errors from the CI run. Could you please fix them?

bganapa commented:
@lebarano Please fix the issues here.. Cant keep this open for long

lebarano committed:
Fixed CI issues and renamed some apis

Fixed CI issues and renamed some apis

bganapa commented:
@lebarano There are still few errors
https://msazure.visualstudio.com/DefaultCollection/One/_git/AzureStack-Solution-Deploy/pullrequest/2168997

You should run autorest --azure-validataor and oav validate-spec <spec_json_filepath> and oav validate-example <spec_json_filepath> locally on your box to validate( the instructions are in our one note as well)

lebarano committed:
Fixed naming issues

Fixed naming issues

lebarano committed:
Ran vs code format on jsons

Ran vs code format on jsons

bganapa commented:
@jhendrixMSFT This should be good to merge finally!. Thanks

jhendrixMSFT commented:
@bganapa  is this functionality live?

lebarano commented:
Some of the APIs updated here are already publically live, some are available but are only used by test environments only: will be delivered to public with next major AzureStack version

bganapa commented:
To re phrase what Leo said, changes in the Microsoft.AzureStack namespaces are live in Azure(They are public cloud changes). The changes in Microsoft.AzureBrdige.Admin namespace is that of specific to AZURESTACK. Those changes gets shipped via monthly updates of AZURESTACK and its due this week. I don't see an issue in merging, we usually merge these azurestack admin api changes closer to the release.

jhendrixMSFT commented:
Merging into the public repo is considered a release event and it's expected that all content is live and that customers can generate functional SDKs from the swagger files.  If there is urgency I can merge this however in the future please keep in mind that the expectation is the service is live before merging.

lebarano committed:
Removed configuration API

Removed configuration API

lebarano commented:
I've remoed configuration api: it was not available for non-test environments yet, and we've made a decision to eliminate it completely in favor of existing ways to get the same data

jhendrixMSFT commented:
Looks like ARM signoff is missing.

bganapa commented:
@ryansbenson  Could you please approve this PR?

bganapa commented:
@ryansbenson  Could you please approve this PR? Its open for a long time

MeBaranov commented:
@ryansbenson Ping

KrisBash reviewed (approved):
Signing off for aRM

bganapa commented:
@jhendrixMSFT ARMSIgnedOff. Please proceed to merge
